contracts/strategies/traderjoe/TraderJoe.sol

```diff
@@ -31,47 +30,135 @@ contract TraderJoe is UUPSUpgradeable, StrategyOwnablePausableBaseUpgradeable {
+        uint256[] calldata binIds,
+        uint256[] calldata binAllocationsX,
+        uint256[] calldata binAllocationsY
     ) external initializer {
+        checkBinIds(binIds);
+        checkBinAllocations(binAllocationsX);
+        checkBinAllocations(binAllocationsY);
+
         __UUPSUpgradeable_init();
         __StrategyOwnablePausableBaseUpgradeable_init(strategyArgs);
 
         TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
             .getStorage();
 
-        strategyStorage.router = router;
-        strategyStorage.masterChef = masterChef;
-        strategyStorage.lpToken = lpToken;
-        strategyStorage.joeToken = joeToken;
-
-        address token0 = lpToken.token0();
-        address token1 = lpToken.token1();
-        if (token0 != address(depositToken)) {
-            strategyStorage.pairDepositToken = IERC20Upgradeable(token0);
-        } else if (token1 != address(depositToken)) {
-            strategyStorage.pairDepositToken = IERC20Upgradeable(token1);
+        strategyStorage.lbPair = lbPair;
+        strategyStorage.lbRouter = lbRouter;
+        strategyStorage.tokenX = IERC20Upgradeable(lbPair.tokenX());
+        strategyStorage.tokenY = IERC20Upgradeable(lbPair.tokenY());
+        strategyStorage.binStep = binStep;
+        strategyStorage.binIds = binIds;
+        strategyStorage.binAllocationsX = binAllocationsX;
+        strategyStorage.binAllocationsY = binAllocationsY;
+
+        if (strategyStorage.tokenX == depositToken) {
+            strategyStorage.pairDepositToken = IERC20Upgradeable(
+                strategyStorage.tokenY
+            );
+        } else if (strategyStorage.tokenY == depositToken) {
+            strategyStorage.pairDepositToken = IERC20Upgradeable(
+                strategyStorage.tokenX
+            );
         } else {
-            revert InvalidTraderJoeLpToken();
+            revert InvalidTraderJoeLBPair();
         }
+    }
 
-        ITraderJoeMasterChef.PoolInfo memory poolInfo;
-        uint256 poolLength = masterChef.poolLength();
-        bool isPoolFound = false;
-        for (uint256 i = 0; i < poolLength; i++) {
-            poolInfo = masterChef.poolInfo(i);
-            if (address(poolInfo.lpToken) == address(lpToken)) {
-                strategyStorage.farmId = i;
-                isPoolFound = true;
-                break;
-            }
-        }
+    function reinitialize(
+        ITraderJoeLBPair lbPair,
+        ITraderJoeLBRouter lbRouter,
+        uint256 binStep,
+        uint256[] calldata binIds,
+        uint256[] calldata binAllocationsX,
+        uint256[] calldata binAllocationsY,
+        NameValuePair[] calldata params
+    ) external reinitializer(2) {
+        checkBinIds(binIds);
+        checkBinAllocations(binAllocationsX);
+        checkBinAllocations(binAllocationsY);
 
-        if (!isPoolFound) {
-            revert InvalidTraderJoeLpToken();
+        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
```

Comment:
maybe you have tested it, but how much money will we lose with our current prod portfolio, if we do this?

Reply:
due to swap fees

Reply:
Please see calculation below.